infra/gravity/cluster_status.go

```diff
@@ -2,45 +2,108 @@ package gravity
 
 import (
 	"context"
-	"time"
 
+	"github.com/cenkalti/backoff"
+
+	"github.com/gravitational/robotest/lib/constants"
 	sshutils "github.com/gravitational/robotest/lib/ssh"
 	"github.com/gravitational/robotest/lib/utils"
 	"github.com/gravitational/robotest/lib/wait"
-
 	"github.com/gravitational/trace"
 )
 
-// Status walks around all nodes and checks whether they all feel OK
-func (c *TestContext) Status(nodes []Gravity) error {
-	c.Logger().WithField("nodes", Nodes(nodes)).Info("Check status on nodes.")
-	ctx, cancel := context.WithTimeout(c.ctx, c.timeouts.Status)
-	defer cancel()
+// returns nil if the Gravity Status is the expected status or an error otherwise
+type statusValidator func(s GravityStatus) error
 
-	retry := wait.Retryer{
-		Attempts: 100,
-		Delay:    time.Second * 20,
+// a negation / reimplementation of https://github.com/gravitational/gravity/blob/7.0.0/lib/status/status.go#L180-L185
+// I'd prefer isDegraded and negation later, but I didn't find an elegant way to
+// have consumer friendly error messages when the information channel is squashed to != nil
+// -- 2020-05 walt
+func isNotDegraded(s GravityStatus) error {
+	if s.Cluster.State == constants.SiteStateDegraded {
+		return trace.Errorf("cluster state %q", s.Cluster.State)
+	}
+	if s.Cluster.SystemStatus != constants.SystemStatus_Running {
+		return trace.Errorf("expected system_status %v, found %v", constants.SystemStatus_Running, s.Cluster.SystemStatus)
 	}
+	return nil
 
-	err := retry.Do(ctx, func() error {
-		errs := make(chan error, len(nodes))
+}
 
-		for _, node := range nodes {
-			go func(n Gravity) {
-				_, err := n.Status(ctx)
-				errs <- err
-			}(node)
-		}
+// not degraded and state == active
+func isActive(s GravityStatus) error {
+	if err := isNotDegraded(s); err != nil {
+		return trace.Wrap(err)
+	}
+	if s.Cluster.State != constants.SiteStateActive {
+		return trace.Errorf("expected state %q, found %q", constants.SiteStateActive, s.Cluster.State)
+	}
+	return nil
+}
+
+func (c *TestContext) WaitForActiveStatus(nodes []Gravity) error {
+	c.Logger().WithField("nodes", Nodes(nodes)).Info("Waiting for active status.")
+	return c.WaitForStatus(nodes, isActive)
+}
+
+// Waits for all nodes to be in the same status, returns an error if nodes do not converge
+func (c *TestContext) WaitForStatus(nodes []Gravity, expected statusValidator) error {
+	b := backoff.NewExponentialBackOff()
+	b.MaxElapsedTime = c.timeouts.ClusterStatus
 
-		err := utils.CollectErrors(ctx, errs)
-		if err == nil {
-			return nil
+	expectStatus := func() (err error) {
+		statuses, err := c.Status(nodes)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		for _, sp := range statuses {
+			if sp == nil { // should never trigger, as the above err != nil check catches this path on well shaped data
```

Comment:
This should never be part of the slice then. I also commented on the use of pointers vs values - we don't need pointers here, and if you used values, there would be no need for `nil` test.

Reply:
Addressed in 28aab700.